Objects/dictobject.c

```diff
@@ -1709,7 +1734,38 @@ insertdict(PyInterpreterState *interp, PyDictObject *mp,
         assert(mp->ma_keys->dk_kind == DICT_KEYS_GENERAL);
     }
 
-    Py_ssize_t ix = _Py_dict_lookup(mp, key, hash, &old_value);
+    Py_ssize_t ix;
+#ifdef Py_GIL_DISABLED
+    PyDictKeysObject *dk = mp->ma_keys;
+    DictKeysKind kind = dk->dk_kind;
+
+    if (kind == DICT_KEYS_SPLIT) {
+        // Split keys can be mutated by other dictionaries, so we need
+        // to do a threadsafe lookup here.  This is basically the split-key
+        // half of _Py_dict_lookup_threadsafe and avoids locking the
+        // shared keys if we can
+        if (PyUnicode_CheckExact(key)) {
+            ix = unicodekeys_lookup_unicode_threadsafe(dk, key, hash);
+        }
+        else {
+            ix = unicodekeys_lookup_generic_threadsafe(mp, dk, key, hash);
```

Comment:
This case shouldn't happen: we have a check `if (DK_IS_UNICODE(mp->ma_keys) && !PyUnicode_CheckExact(key))` above that should convert a split keys into a unified if `key` is not `PyUnicode_CheckExact`.

I think this should just be an assertion `assert(PyUnicode_CheckExact(key));`